# Lending Club Risk Prediction

The aim of this project is to build a machine learning model that will allow us to accurately predict if a borrower will pay off their loan on time or not. 

The dataset we'll be using is the complete loan data for all [Lending Club](https://www.lendingclub.com) loans issued from 2012 to 2013, which can be downloaded [here](https://www.lendingclub.com/info/download-data.action).

A description of each column can be found [here](https://docs.google.com/spreadsheets/d/191B2yJ4H1ZPXq0_ByhUgWMFZOYem5jFz0Y3by_7YBY4/edit#gid=2081333097).

# Importing and Cleaning the Data

In [1]:
import pandas as pd
import numpy as np

# We'll skip the first row because it provides information about the dataset's origin rather than column titles.
loans_2013 = pd.read_csv('loans2013.csv', skiprows=1)

og_cols = loans_2013.shape[1]
# Drop columns in which at least half of the column data is missing
missing_info_thresh = loans_2013.shape[0]/2
loans_2013.dropna(thresh=missing_info_thresh, axis=1, inplace=True)

# Save processed dataset incase we want to explore the original later
loans_2013.to_csv('processed_loans2013.csv', index=False)

print('{} columns dropped'.format(og_cols - loans_2013.shape[1]))

loans_2013.head(1)

/home/zrupp/anaconda3/lib/python3.7/site-packages/IPython/core/interactiveshell.py:3020: DtypeWarning: Columns (0,47,123,124,125,128,129,130,133) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


58 columns dropped


loan_amnt  funded_amnt  funded_amnt_inv        term int_rate  installment  \
0    12000.0      12000.0          12000.0   36 months    7.62%       373.94   

  grade sub_grade         emp_title emp_length         ...           \
0     A        A3  Systems Engineer    3 years         ...            

  percent_bc_gt_75  pub_rec_bankruptcies tax_liens tot_hi_cred_lim  \
0            100.0                   0.0       0.0        233004.0   

  total_bal_ex_mort total_bc_limit total_il_high_credit_limit hardship_flag  \
0           46738.0        14800.0                    53404.0             N   

  disbursement_method debt_settlement_flag  
0                Cash                    N  

[1 rows x 87 columns]

That's a lot of columns. Determining which columns to keep is going to take some time. Let's familiarize ourselves with the columns and determine which columns:

- Leak information from the future
- Don't affect a borrower's ability to pay back a loan
- Need to be cleaned because of poor formatting
- Require more data or too much processing to be usefully turned into a feature
- Contain redundant information

We'll want to import the data dictionary to make this process a little less painful.

# Columns 1-20

In [2]:
# We'll need to expand the maximum column width to view the descriptions
pd.options.display.max_colwidth = 250

# Import the data dictionary to easily reference what each column describes
data_dict = pd.read_excel('LCDataDictionary.xlsx', sheet_name='LoanStats')

data_dict.head()

LoanStatNew  \
0        acc_now_delinq   
1  acc_open_past_24mths   
2            addr_state   
3              all_util   
4            annual_inc   

                                                                     Description  
0                The number of accounts on which the borrower is now delinquent.  
1                                     Number of trades opened in past 24 months.  
2                     The state provided by the borrower in the loan application  
3                                          Balance to credit limit on all trades  
4  The self-reported annual income provided by the borrower during registration.

In [3]:
def explain_columns(start, end=None):
    '''
    Returns a dataframe with index, column name, description of column, and first value in column
    
    start: starting index
    end: ending index 
    '''

    first_n_cols = loans_2013.columns[start:end]
    indices = []
    first_n_vals = pd.Series()


    for col in first_n_cols:
        index = list(data_dict[data_dict['LoanStatNew'] == col].index)
        indices.extend(index)
        first_n_vals = first_n_vals.append(pd.Series(loans_2013[col][0], index))

    first_n = data_dict[data_dict.index.isin(indices)]
    
    

    return pd.concat([first_n, pd.DataFrame(data=first_n_vals, columns=['first value'])], axis=1).reset_index(drop=True)


explain_columns(0, 20)

LoanStatNew  \
0            addr_state   
1            annual_inc   
2            emp_length   
3             emp_title   
4           funded_amnt   
5       funded_amnt_inv   
6                 grade   
7        home_ownership   
8           installment   
9              int_rate   
10              issue_d   
11            loan_amnt   
12          loan_status   
13              purpose   
14           pymnt_plan   
15            sub_grade   
16                 term   
17                title   
18  verification_status   
19             zip_code   

                                                                                                                                                                     Description  \
0                                                                                                                     The state provided by the borrower in the loan application   
1                                                                                                  The self-reported annual income provided by the borrower during registration.   
2                                             Employment length in years. Possible values are between 0 and 10 where 0 means less than one year and 10 means ten or more years.    
3                                                                                                            The job title supplied by the Borrower when applying for the loan.*   
4                                                                                                                 The total amount committed to that loan at that point in time.   
5                                                                                                   The total amount committed by investors for that loan at that point in time.   
6                                                                                                                                                         LC assigned loan grade   
7                          The home ownership status provided by the borrower during registration or obtained from the credit report. Our values are: RENT, OWN, MORTGAGE, OTHER   
8                                                                                                               The monthly payment owed by the borrower if the loan originates.   
9                                                                                                                                                      Interest Rate on the loan   
10                                                                                                                                           The month which the loan was funded   
11  The listed amount of the loan applied for by the borrower. If at some point in time, the credit department reduces the loan amount, then it will be reflected in this value.   
12                                                                                                                                                    Current status of the loan   
13                                                                                                                    A category provided by the borrower for the loan request.    
14                                                                                                                Indicates if a payment plan has been put in place for the loan   
15                                                                                                                                                     LC assigned loan subgrade   
16                                                                                          The number of payments on the loan. Values are in months and can be either 36 or 60.   
17                                                                                                                                       The loan title provided by the borrower   
18                      

In [4]:
# It may be meaningless, but let's find out what that asterisk signifies
data_dict.iloc[-1:]

LoanStatNew  \
152         NaN   

                                                                      Description  
152  * Employer Title replaces Employer Name for all loans listed after 9/23/2013

Columns to drop:

- emp_title: relevance is suspect. Also likely contains a ton of different possible categorical values which means turning this into a useful feature will take too much processing to be useful
- funded_amnt: leaks future data
- funded_amnt_inv: leaks future data
- grade: basically a weaker version of int_rate
- issue_d: leaks future data
- sub_grade: same as grade
- zip_code: basically the same as addr_state

In [5]:
dropped_cols = ['emp_title', 'funded_amnt', 'funded_amnt_inv', 'grade', 'issue_d', 'sub_grade', 'zip_code']
len(dropped_cols)

7

# Columns 21-40

In [6]:
explain_columns(20, 40)

LoanStatNew  \
0   collection_recovery_fee   
1               delinq_2yrs   
2                       dti   
3          earliest_cr_line   
4       initial_list_status   
5            inq_last_6mths   
6              last_pymnt_d   
7                  open_acc   
8                 out_prncp   
9             out_prncp_inv   
10                  pub_rec   
11               recoveries   
12                revol_bal   
13               revol_util   
14                total_acc   
15              total_pymnt   
16          total_pymnt_inv   
17            total_rec_int   
18       total_rec_late_fee   
19          total_rec_prncp   

                                                                                                                                                                                                 Description  \
0                                                                                                                                                                             post charge off collection fee   
1                                                                                               The number of 30+ days past-due incidences of delinquency in the borrower's credit file for the past 2 years   
2   A ratio calculated using the borrower’s total monthly debt payments on the total debt obligations, excluding mortgage and the requested LC loan, divided by the borrower’s self-reported monthly income.   
3                                                                                                                                          The month the borrower's earliest reported credit line was opened   
4                                                                                                                                         The initial listing status of the loan. Possible values are – W, F   
5                                                                                                                           The number of inquiries in past 6 months (excluding auto and mortgage inquiries)   
6                                                                                                                                                                            Last month payment was received   
7                                                                                                                                             The number of open credit lines in the borrower's credit file.   
8                                                                                                                                                    Remaining outstanding principal for total amount funded   
9                                                                                                                            Remaining outstanding principal for portion of total amount funded by investors   
10                                                                                                                                                                       Number of derogatory public records   
11                                                                                                                                                                            post charge off gross recovery   
12                                                                                                                                                                            Total credit revolving balance   
13                                                                                Revolving line utilization rate, or the amount of credit the borrower is using relative to all available revolving credit.   
14                                                                                                                                  The total number of credit lines currently in the borrower's credit file   
15                                  

Columns to drop:
- collection_recovery_fee: leaks future data
- last_pymnt_d: leaks future data
- out_prncp: leaks future data
- out_prncp_inv: same as out_prncp
- recoveries: leaks future data
- total_pymnt: leaks future data
- total_pymnt_inv: same as total_pymnt
- total_rec_int: leaks future data
- total_rec_late_fee: leaks future data
- total_rec_prncp: leaks future data


In [7]:
dropped_cols.extend([
    'collection_recovery_fee', 
    'last_pymnt_d',
    'out_prncp',
    'out_prncp_inv',
    'recoveries',
    'total_pymnt',
    'total_pymnt_inv',
    'total_rec_int',
    'total_rec_late_fee',
    'total_rec_prncp'
])

len(dropped_cols)

17

# Columns 41-60

In [8]:
explain_columns(40, 60)

LoanStatNew  \
0               acc_now_delinq   
1         acc_open_past_24mths   
2             application_type   
3                  avg_cur_bal   
4               bc_open_to_buy   
5                      bc_util   
6     chargeoff_within_12_mths   
7   collections_12_mths_ex_med   
8                  delinq_amnt   
9           last_credit_pull_d   
10             last_pymnt_amnt   
11          mo_sin_old_il_acct   
12        mo_sin_old_rev_tl_op   
13       mo_sin_rcnt_rev_tl_op   
14              mo_sin_rcnt_tl   
15                    mort_acc   
16                 policy_code   
17                tot_coll_amt   
18                 tot_cur_bal   

                                                                                             Description  \
0                                        The number of accounts on which the borrower is now delinquent.   
1                                                             Number of trades opened in past 24 months.   
2   Indicates whether the loan is an individual application or a joint application with two co-borrowers   
3                                                                Average current balance of all accounts   
4                                                              Total open to buy on revolving bankcards.   
5                  Ratio of total current balance to high credit/credit limit for all bankcard accounts.   
6                                                                 Number of charge-offs within 12 months   
7                                       Number of collections in 12 months excluding medical collections   
8                     The past-due amount owed for the accounts on which the borrower is now delinquent.   
9                                                   The most recent month LC pulled credit for this loan   
10                                                                    Last total payment amount received   
11                                                   Months since oldest bank installment account opened   
12                                                          Months since oldest revolving account opened   
13                                                     Months since most recent revolving account opened   
14                                                               Months since most recent account opened   
15                                                                          Number of mortgage accounts.   
16                   publicly available policy_code=1\nnew products not publicly available policy_code=2   
17                                                                    Total collection amounts ever owed   
18                                                                 Total current balance of all accounts   

   first value  
0            0  
1            4  
2   Individual  
3        11783  
4         2441  
5         83.5  
6            0  
7            0  
8            0  
9     Mar-2019  
10     2927.22  
11         123  
12         118  
13          10  
14           9  
15           1  
16           1  
17           0  
18      200314

Columns to drop:
- last_pymnt_amnt: leaks future data (after loan start date)


In [9]:
dropped_cols.extend([
    'last_pymnt_amnt',
])

len(dropped_cols)

18

# Columns 61-80

In [10]:
explain_columns(60,80)

LoanStatNew  \
0    mths_since_recent_bc   
1   mths_since_recent_inq   
2   num_accts_ever_120_pd   
3          num_actv_bc_tl   
4         num_actv_rev_tl   
5             num_bc_sats   
6               num_bc_tl   
7               num_il_tl   
8           num_op_rev_tl   
9           num_rev_accts   
10    num_rev_tl_bal_gt_0   
11               num_sats   
12       num_tl_120dpd_2m   
13           num_tl_30dpd   
14     num_tl_90g_dpd_24m   
15     num_tl_op_past_12m   
16         pct_tl_nvr_dlq   
17       percent_bc_gt_75   
18   pub_rec_bankruptcies   
19              tax_liens   

                                                                  Description  \
0                           Months since most recent bankcard account opened.   
1                                           Months since most recent inquiry.   
2                           Number of accounts ever 120 or more days past due   
3                                Number of currently active bankcard accounts   
4                                 Number of currently active revolving trades   
5                                    Number of satisfactory bankcard accounts   
6                                                 Number of bankcard accounts   
7                                              Number of installment accounts   
8                                           Number of open revolving accounts   
9                                                Number of revolving accounts   
10                                 Number of revolving trades with balance >0   
11                                            Number of satisfactory accounts   
12  Number of accounts currently 120 days past due (updated in past 2 months)   
13   Number of accounts currently 30 days past due (updated in past 2 months)   
14              Number of accounts 90 or more days past due in last 24 months   
15                                Number of accounts opened in past 12 months   
16                                         Percent of trades never delinquent   
17                        Percentage of all bankcard accounts > 75% of limit.   
18                                       Number of public record bankruptcies   
19                                                        Number of tax liens   

    first value  
0          10.0  
1          10.0  
2           0.0  
3           4.0  
4           5.0  
5           4.0  
6          10.0  
7          15.0  
8           8.0  
9          14.0  
10          5.0  
11         17.0  
12          0.0  
13          0.0  
14          0.0  
15          3.0  
16        100.0  
17        100.0  
18          0.0  
19          0.0

None of these columns seem problematic.

# Remaining Columns

In [11]:
explain_columns(80)

LoanStatNew  \
0             tot_hi_cred_lim   
1           total_bal_ex_mort   
2              total_bc_limit   
3  total_il_high_credit_limit   
4               hardship_flag   
5         disbursement_method   
6        debt_settlement_flag   

                                                                                          Description  \
0                                                                      Total high credit/credit limit   
1                                                             Total credit balance excluding mortgage   
2                                                             Total bankcard high credit/credit limit   
3                                                          Total installment high credit/credit limit   
4                                             Flags whether or not the borrower is on a hardship plan   
5         The method by which the borrower receives their loan. Possible values are: CASH, DIRECT_PAY   
6  Flags whether or not the borrower, who has charged-off, is working with a debt-settlement company.   

  first value  
0      233004  
1       46738  
2       14800  
3       53404  
4           N  
5        Cash  
6           N

These all look fine. On reflection, there are two columns I'm not entirely sure about: total_il_high_credit_limit, and num_il_tl. Because Lending Club seems to only offer installment accounts, it's possible both of these columns leak future data. We'll leave them for now, but we may want to try dropping these columns if we notice overfitting.

In [12]:
loans_2013.drop(dropped_cols, axis=1, inplace=True)
loans_2013.head()

loan_amnt        term int_rate  installment emp_length home_ownership  \
0    12000.0   36 months    7.62%       373.94    3 years       MORTGAGE   
1    27050.0   36 months   10.99%       885.46  10+ years            OWN   
2    12000.0   36 months   11.99%       398.52  10+ years       MORTGAGE   
3    28000.0   36 months    7.62%       872.52    5 years       MORTGAGE   
4    12000.0   36 months   10.99%       392.81    4 years           RENT   

   annual_inc verification_status loan_status pymnt_plan  \
0     96500.0        Not Verified  Fully Paid          n   
1     55000.0            Verified  Fully Paid          n   
2    130000.0     Source Verified  Fully Paid          n   
3    325000.0     Source Verified  Fully Paid          n   
4     60000.0        Not Verified  Fully Paid          n   

           ...          percent_bc_gt_75 pub_rec_bankruptcies tax_liens  \
0          ...                     100.0                  0.0       0.0   
1          ...                      25.0                  0.0       0.0   
2          ...                       1.0                  0.0       0.0   
3          ...                      16.7                  0.0       0.0   
4          ...                       0.0                  0.0       0.0   

   tot_hi_cred_lim  total_bal_ex_mort total_bc_limit  \
0         233004.0            46738.0        14800.0   
1         138554.0            70186.0        35700.0   
2         365874.0            44327.0        10700.0   
3         850886.0           199739.0        42200.0   
4          29700.0             7137.0        18100.0   

   total_il_high_credit_limit  hardship_flag  disbursement_method  \
0                     53404.0              N                 Cash   
1                     33054.0              N                 Cash   
2                     57674.0              N                 Cash   
3                    196686.0              N                 Cash   
4                         0.0              N                 Cash   

   debt_settlement_flag  
0                     N  
1                     N  
2                     N  
3                     N  
4                     N  

[5 rows x 69 columns]

69 columns may not seem like much, but I want to reduce dimensionality as much as possible before I start transforming the data, so let's see if we can drop some more columns with high null counts.

# Dropping Columns with High Null Counts

In [13]:
null_counts = loans_2013.isnull().sum()

null_counts[null_counts/loans_2013.shape[0] > .05]/loans_2013.shape[0]

tot_coll_amt                  0.147426
tot_cur_bal                   0.147426
total_rev_hi_lim              0.147426
avg_cur_bal                   0.147458
mo_sin_old_il_acct            0.180006
mo_sin_old_rev_tl_op          0.147431
mo_sin_rcnt_rev_tl_op         0.147431
mo_sin_rcnt_tl                0.147426
mths_since_recent_inq         0.148101
num_accts_ever_120_pd         0.147426
num_actv_bc_tl                0.147426
num_actv_rev_tl               0.147426
num_bc_sats                   0.085327
num_bc_tl                     0.147426
num_il_tl                     0.147426
num_op_rev_tl                 0.147426
num_rev_accts                 0.147426
num_rev_tl_bal_gt_0           0.147426
num_sats                      0.085327
num_tl_120dpd_2m              0.148786
num_tl_30dpd                  0.147426
num_tl_90g_dpd_24m            0.147426
num_tl_op_past_12m            0.147426
pct_tl_nvr_dlq                0.148239
tot_hi_cred_lim               0.147426
total_il_high_credit_limi

A significant number of columns are missing approximately 15% of the data. We could handle this a few ways:

1. We could just drop the columns entirely
2. We could drop the rows with the missing data
3. We could replace missing values with the most common value of the relevant columns.
4. We could replace missing values with 0s

I'm inclined to drop the columns entirely, but looking at these columns it seems this information may in fact be valuable. I don't want to drop the rows because that's a substantial amount of potentially meaningful data we will lose. Replacing missing values in these columns with the most common values or 0s seems like a huge mess because we're inserting data which may be extremely removed from reality, which could potentially throw off our model. It seems the best course of action would be to drop the columns entirely.